src/camera.c

```diff
@@ -6,43 +6,50 @@
 /*   By: gwolf <gwolf@student.42vienna.com>         +#+  +:+       +#+        */
 /*                                                +#+#+#+#+#+   +#+           */
 /*   Created: 2023/10/28 11:48:56 by u413q             #+#    #+#             */
-/*   Updated: 2023/12/22 12:02:39 by gwolf            ###   ########.fr       */
+/*   Updated: 2023/12/25 08:47:19 by gwolf            ###   ########.fr       */
 /*                                                                            */
 /* ************************************************************************** */
 
 #include "camera.h"
 
-void	ft_initiate_camera(t_cam *cam)
+void	ft_cam_init(t_cam *cam, t_vec2i img_size)
 {
-	cam->vup.x = 0.0;
-	cam->vup.y = 1.0;
-	cam->vup.z = 0.0;
-	cam->camera_centre.x = cam->look_from.x;
-	cam->camera_centre.y = cam->look_from.y;
-	cam->camera_centre.z = cam->look_from.z;
-	cam->w = ft_vec3_norm(ft_vec3_sub(cam->look_from, cam->look_at));
-	cam->u = ft_vec3_cross(cam->vup, cam->w);
+	ft_cam_calc_base_vectors(cam);
+	cam->focal_length = 10.0;
+	cam->image = img_size;
+	ft_cam_calc_viewport_dimensions(cam);
+	ft_cam_calc_pixel_grid(cam);
+}
+
+void	ft_cam_calc_base_vectors(t_cam *cam)
+{
+	const t_vec3	world_up = {0.0, 1.0, 0.0};
+
+	cam->w = ft_vec3_norm(ft_vec3_scale(cam->direction, -1));
```

Comment:
That's how the base vectors are calculated after investigating different sources and repos:
	
cam->vup = (t_vec3){0.0, 1.0, 0.0};
	cam->camera_centre = cam->look_from;
	cam->look_at = ft_vec3_add(cam->camera_centre, cam->camera_direction);
	cam->w = ft_vec3_norm(ft_vec3_sub(cam->look_from, cam->look_at));
	if (cam->camera_direction.y == 1)
		cam->u = (t_vec3){1, 0, 0};
	else if (cam->camera_direction.y == -1)
		cam->u = (t_vec3){-1, 0, 0};
	else
		cam->u = ft_vec3_norm(ft_vec3_cross(cam->vup, cam->w));
	cam->v = ft_vec3_cross(cam->w, cam->u);

This is used currently in my code.

Reply:
camera_direction is the orientation vector parsed from file

Reply:
Some of these operations cancel each other out.

We want to calculate the local coordinate system of the camera, which are three orthonormal vectors - essentially x, y and z of the camera, or direction, up and right. We are fancy, so we call them u, v, w. 

Right now it is calculated as:
```
1 const t_vec3	world_up = {0.0, 1.0, 0.0};

2 cam->w = ft_vec3_norm(ft_vec3_scale(cam->direction, -1));
3 cam->u = ft_vec3_cross(world_up, cam->w);
4 cam->v = ft_vec3_cross(cam->w, cam->u);
```
I would even get rid of the flipping, since I don't understand why we do it. This would mean we parse directly into cam.w.

The code above does the same thing, except that it guards against a special case (good) and does some not needed operations (bad). Simplified:

- Define world_up as (0,1,0)
- A = A > Rename the camera center vector to look_from. I think we should use one nomenclature and stick to it, not mix around. This could also mask some stuff which is happening (as demonstrated right here)
- C = A + B > Calc the look_at vector. Since it is adding, we at least don't run into the (0,0,0) issue.
- w = A - C == A - (A + B) == -B > we did some calculations to set w to flipped direction
- Check if cam is looking exactly up or down. In this case we would have a problem since 
(0,1,0) x (0,1,0) == world_up x cam_direction = (0,0,0), which is no unit vector
- Cross cam_direction and cam_up for cam_right

Line 1 saves the world_up vector in cam struct. We generally only need it when calc cam_up, which is why I have it as static const in line 1 of the function.
Lines 2-4 are obfuscating what is happening, they are essentially the same as 2.
The protection against world_up == direction is good. I will add that.
The last line is the same.

Reply:
Let's stick to look_from and look_at, these seem to be commonly accepted terms in the raytracing field.

The vup vector can/should never be the same as w. The cases when the cam is looking up or down (cam->direction = 1 or -1) are covered by the extra cases. Any other position of the camera is then calculated with the cross product.

We can leave vup as a const.

Reply:
> Let's stick to look_from and look_at, these seem to be commonly accepted terms in the raytracing field.

Regarding naming? What about the (in my view needless) calculations? 

> The vup vector can/should never be the same as w. The cases when the cam is looking up or down (cam->direction = 1 or -1) are covered by the extra cases. Any other position of the camera is then calculated with the cross product.

Yes it can be the same, which is why we need the special cases.

Reply:
Yes naming. The calculations can be discussed in person if the writing isn't clear enough.

Reply:
But we are on the same page that your suggested version produces the same result in a more roundabout way?

Reply:
It seems that we get the same results as far as I can tell from a distance.